### PR TITLE
Feature: Add Campaign model and repository

### PR DESCRIPTION
**Pull request by alaca**
## Description

This PR adds a Campaign model and repository.

## Pre-review Checklist

-   [x] Relevant `@unreleased` tags included in DocBlocks
-   [x] Includes unit tests
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

### CODE REVIEW

File: src/Campaigns/Repositories/CampaignRepository.php [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        DB::query('START TRANSACTION');
+
+        Hooks::doAction('givewp_campaign_deleting', $campaign);
+
+        try {
+            DB::table('give_campaigns')
+                ->where('id', $campaign->id)
+                ->delete();
+
+            // todo: also delete campaign page?
```
Comment (kjohnson) [inline src/Campaigns/Repositories/CampaignRepository.php:148]: We'll probably want to hook into the `givewp_campaign_deleting` hook above for the campaign page to delete itself (which or may not exist).

File: src/Campaigns/ValueObjects/CampaignType.php [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * @method static CampaignStatus CORE()
+ * @method static CampaignStatus P2P()
+ * @method bool isCore()
+ * @method bool isP2p()
+ */
+class CampaignType extends Enum
+{
+    const CORE = 'core';
+    const P2P = 'p2p';
```
Comment (kjohnson) [inline src/Campaigns/ValueObjects/CampaignType.php:18]: Should we start referring to P2P using the pull name, like `const PEER_TO_PEER` and `isPeerToPeer()`? Now that we are moving into a larger context, I don't want shorthand and abbreviation to create ambiguity.

Comment (alaca) [reply src/Campaigns/ValueObjects/CampaignType.php]: Good point! Resolved https://github.com/impress-org/givewp/pull/7520/commits/a72f80c29861921f5c493dfbf22a550abf3e7e70

File: src/Campaigns/Repositories/CampaignRepository.php (hunk 1 above)
Comment (alaca) [reply src/Campaigns/Repositories/CampaignRepository.php:148]: We can create an action later, for now, I will just update the todo tag.

Comment (alaca) [reply src/Campaigns/Repositories/CampaignRepository.php]: Resolved: https://github.com/impress-org/givewp/pull/7520/commits/10be69b385c1bc4a593dd8af60020915463c5762

**Review by kjohnson — APPROVED** [review]
LGTM LFG

**Pull request merged**